velox/exec/MergingVectorInput.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/MergingVectorInput.h"
+#include "velox/exec/OperatorUtils.h"
+
+namespace facebook::velox::exec {
+
+MergingVectorInput::MergingVectorInput(
+    velox::memory::MemoryPool* pool,
+    int64_t preferredOutputBatchBytes,
+    int32_t preferredOutputBatchRows,
+    int32_t maxMergingRows,
+    bool preserveOrder)
+    : pool_(pool),
+      preferredOutputBatchBytes_(preferredOutputBatchBytes),
+      preferredOutputBatchRows_(preferredOutputBatchRows),
+      maxMergingRows_(maxMergingRows),
+      preserveOrder_(preserveOrder) {}
+
+RowVectorPtr MergingVectorInput::getVector(bool noMoreInput) {
+  if (noMoreInput) {
+    flush();
+  }
+  if (outputQueue_.size() > 0) {
+    const auto res = outputQueue_.front();
+    outputQueue_.pop();
+    return res;
+  }
+  return nullptr;
+}
+
+void MergingVectorInput::addVector(RowVectorPtr vector) {
+  if (nullptr == vector) {
+    return;
+  }
+
+  const auto numInput = vector->size();
+  if (numInput == 0) {
+    return;
+  }
+
+  // Avoid memory copying for pages that are big enough
+  if (numInput >= maxMergingRows_) {
+    if (preserveOrder_) {
+      // In order to preserve the order of the rows, we need to flush first.
+      flush();
+    }
+    outputQueue_.push(std::move(vector));
+    return;
+  }
+
+  buffer(std::move(vector));
+}
+
+void MergingVectorInput::buffer(RowVectorPtr vector) {
+  const auto numInput = vector->size();
+
+  // In order to prevent vector.resize() cause memory copy, the initial value
```

Comment:
Not sure if this comment is relevant. Is it to prevent we buffer more than preferredOutputBatchRows_? Thanks!

Reply:
Yes, the buffer init size is preferredOutputBatchRows_, if the buffer more than preferredOutputBatchRows_ will cause memory copy.